core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java

```diff
@@ -251,6 +257,26 @@ public Expression rewriteFunctionCall(FunctionCall node, Void context, Expressio
                 }
             }
 
+            if (functionName.equals("date_trunc")
+                    && node.getArguments().size() == 2
+                    && getType(node.getArguments().get(0)) instanceof VarcharType
+                    && isConstant(node.getArguments().get(0))) {
+                Slice unitValue = (Slice) new ExpressionInterpreter(node.getArguments().get(0), plannerContext, session, expressionTypes)
+                        .optimize(NoOpSymbolResolver.INSTANCE);
+                Expression argument = node.getArguments().get(1);
+                Type argumentType = getType(argument);
+                if ("day".equals(unitValue.toStringUtf8().toLowerCase(Locale.ENGLISH))) {
+                    if (argumentType == DATE) {
+                        // date_trunc(day, a_date) is a no-op
+                        return treeRewriter.rewrite(argument, context);
+                    }
+                    if (argumentType instanceof TimestampType) {
+                        // prefer `CAST(CAST(x as DATE) AS timestamp)` to `date_trunc('day', x)`
```

Comment:
It would be nice to have a link towards `io.trino.sql.planner.iterative.rule.UnwrapCastInComparison` here.